Fix torch.distributed.run init connect timeout by comparing `host` with the current IP list

cz-37 opened the pull request:
Summary:
Pull Request: https://github.com/pytorch/pytorch/issues/79388

Fix torch.distributed.run init connect timeout by comparing `host` with the current IP list.

Test Plan: unit tests

Differential Revision: D41373962

d4l3k reviewed (approved):
LGTM w/ unit tests from internal

cz-37 committed:
Fix torch.distributed.run init connect timeout by comparing `host` with the current IP list (#90221)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/90221

Pull Request: https://github.com/pytorch/pytorch/issues/79388

Fix torch.distributed.run init connect timeout by comparing `host` with the current IP list.

Test Plan:
```
> buck2 test mode/dev-nosan //caffe2/test/distributed/elastic/rendezvous:utils_test -- --exact 'caffe2/test/distributed/elastic/rendezvous:utils_test - test_matches_machine_hostname_returns_true_if_ip_address_match_between_hosts (utils_test.UtilsTest)'

Tests finished: Pass 1. Fail 0. Fatal 0. Skip 0. 0 builds failed
```
Unit tests

Reviewed By: d4l3k

Differential Revision: D41373962

fbshipit-source-id: 7f138e2ef74b057f70271d32b605710bc5d287f6
